### ISSUE TITLE
Fix unstable integration tests

### ISSUE THREAD

**Issue by flomotlik** [serverless/serverless#2655]
# This is a Feature Proposal

## Description

Some integration tests are failing regularly, failing the build on master and are therefore not really trustworthy. We have to fix those to make sure our master is always green. Tests that have failed recently:

* AWS - SNS: Multiple topics with multiple functions should trigger function when new message is published
* AWS - API Gateway (Integration: Lambda): API keys test should succeed if correct API key is given

From my experience this looks suspiciously like a timing issues where we update a specific resource, trigger something and then immediately check if its true without giving it some time until the data could propagate through their system.

Best case going forward would be to implement this with a helper that runs the same check again with same backoff so we can see if the system eventually gets to the right state and the test passes. Another option would be to include a sleep, but thats adding additional instability to it.

Similar or dependent issues:
*

**Comment by gburch** [comment]
I have seen failures on slower internet connections with gradle downloading and downloading dependencies and a key expiring because it took more than 5 minutes. Could be related. One suggestion is to mount a share in docker for gradle and/or the cache of dependencies so that after first run its much faster. Shouldn't effect the outcome of the tests.

**Comment by flomotlik** [comment]
In our specific case as the tests are pure Node and the keys are long lived this shouldn't be the issue, but I've run into the issue you've described often with customers at Codeship. This is super common. Making integration tests stable is a bit of science and art.

**Comment by gburch** [comment]
Here is one of the errors I get, is this the one you are talking about?

`Total time: 31.124 secs
Deploying Service

  Error --------------------------------------------------

     ServerlessError: ServerlessError: Signature expired:
     20161107T202028Z is now earlier than 20161107T202554Z
     (20161107T203054Z - 5 min.)

     For debugging logs, run again after setting SLS_DEBUG env var.

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues

     Please report this error. We think it might be a bug.

  Your Environment Information -----------------------------
     OS:                 linux
     Node Version:       7.0.0
     Serverless Version: 1.1.0`

**Comment by flomotlik** [comment]
@gburch Are you using Docker on OSX? Because this happens regularly when the docker daemon doesn't have its time synced properly (or if your local time isn't properly synced). Restarting the docker daemon will fix this (happens all the time on my machine as well, especially after sleep/resume)

The main issues that make our integration tests unstable are an SNS integration test and an API Gateway test listed above in the issue description.

**Comment by gburch** [comment]
Yes that is the issue. Thanks for the tip. I had thought it was because of long running tests. Sorry for the noise.

**Comment by pmuens** [comment]
Quick status update:

I added the `cfLogs: true` to the `simple-integration-test` recently. The SNS tests in the `complex-integration-test` suite seems to have failed because of the lack of adding log groups (as they were already taken by the `simple-integration-test`). This should resolve the failing SNS test (I tried to reproduce it on my machine with my clean account and it was not popping up).

Currently looking into the API Gateway bug (might be related to the several issues with the inconsistent "usage plans").
